v2/devdoc/log_sink_etw_requirements.md

```diff
@@ -20,160 +20,184 @@ The signature of `log_sink_etw.log_sink_log` is:
 typedef void (*LOG_SINK_LOG_FUNC)(LOG_LEVEL log_level, LOG_CONTEXT_HANDLE log_context, const char* file, const char* func, int line, const char* message_format, ...);
 ```
 
-`log_sink_etw.log_sink_log` implements logging to an ETW provider.
+`log_sink_etw.log_sink_log` implements logging as an ETW provider.
 
-If `message_format` is `NULL`, `log_sink_etw.log_sink_log` shall fail and return.
+**SRS_LOG_SINK_ETW_01_001: [** If `message_format` is `NULL`, `log_sink_etw.log_sink_log` shall return. **]**
 
-`log_sink_etw_log` shall register the ETW TraceLogging provider by calling `TraceLoggingRegister` if it was not already done.
+**SRS_LOG_SINK_ETW_01_002: [** `log_sink_etw_log` shall maintain the state of whether `TraceLoggingRegister` was called in a variable accessed via `InterlockedXXX` APIs, which shall have 3 possible values: `NOT_REGISTERED` (1), `REGISTERING` (2), `REGISTERED`(3). **]**
 
-`log_sink_etw_log` shall use as provider GUID `DAD29F36-0A48-4DEF-9D50-8EF9036B92B4`.
+**SRS_LOG_SINK_ETW_01_003: [** `log_sink_etw_log` shall perform the below actions until the provider is registered or an error is encountered: **]**
 
-Note this can (and should) be improved to be configurable later.
+**SRS_LOG_SINK_ETW_01_004: [** If the state is `NOT_REGISTERED` (1): **]**
 
-`log_sink_etw_log` shall emit a self described event that shall have the name of the event as follows:
+- **SRS_LOG_SINK_ETW_01_005: [** `log_sink_etw_log` shall switch the state to `REGISTERING` (2). **]**
```

Comment:
>shall switch the state to `REGISTERING` (2 [](http://example.com/codeflow?start=52&length=42)

One does have to wonder "if we have platform_init"... why isn't this part registering part of platform_init.

... would save some "ifs" and some "interlocked" for every log that we do.

./consider for later.

Reply:
Yeah that would be awesome.
ID Work Item Type Title State Assigned To Effort Remaining Work Original Estimate Value Area Iteration Path Tags
22152024 Task Initialize logging as part of platform_init to avoid interlocked calls To Do           One

Reply:
I had the same observation some time back for gballoc coming from some profile sessions.